### PR TITLE
Fix unit tests failing on windows

### LINKED ISSUE #3: platform-specific tests are causing unit tests to fail on Windows
## Issue Report
Unit tests are failing on a clean Windows master build.

## Expected Behavior

After a clean clone, yarn, and build, running unit tests should pass on any platform.

## Actual Behavior

On my windows machine, the unit tests are failing.  It seems to have to do with platform-specific paths.

## Steps to Reproduce the Issue

1. Checkout the master branch 
2. Run yarn install 
3. run yarn build 
4. Run yarn test runner-babel or any unit test that deals with paths

## Attempted failed solution

I tried installing jest-serializer-path , but it does not seem that this would solve the issue.  I would need a Mac computer to try.
Also, the last publish from this library was 2 years ago so I'm not sure about its quality.

**Before jest-serializer-path**:

image

**After jest-serializer-path**:

image

If someone could try jest-serializer-path on a Mac that would be great, but I don't see it turning the path I received using it, to those that were snapshotted.

### LINKED ISSUE #7: configFile option does not support npm module syntax
## Issue Report

`{{projectDir}}` is supported, however, npm module syntax ie, `@some-scope/tsconfig/tsconfig.node.json` is causing the error 
described here https://github.com/Farfetch/garment/issues/4 . 

## Expected Behavior

npm module syntax ie, `@some-scope/tsconfig/tsconfig.node.json` should be supported as tsc supports this in tsconfig.json files.

## Actual Behavior

You get this error: 

 image (9)

## Steps to Reproduce the Issue

1. Use npm module syntax either directly in the configFile option of the ts-runner or in the `extends` property of one of the tsconfig files.
2. Run the ts-runner

### PR DESCRIPTION
**Pull request by stephan-noel**
# Description

Fixes #7 This is a partial fix, it won't support npm module syntax in garment.json, only as the extends option in your tsconfig files.
Fixes #3 

No relevant documentation changes to be made.

## How Has This Been Tested?

Most changes were in unit tests.  Source changes (only runner-ts) were also verified manually by running the new code in a standalone project using garment.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have added tests to cover my changes
- [x] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: plugins/runner-ts/__tests__/default.test.ts [hunk 1]
```diff
@@ -85,3 +85,31 @@ test('Transpiles TypeScript', async () => {
 
   expect(result).toMatchSnapshot();
 });
+
+
+test('Transpiles TypeScript when configFile contains npm module in extends field', async () => {
```
Comment (stephan-noel) [inline plugins/runner-ts/__tests__/default.test.ts:89]: This test actuall fails.  I am currently trying some suggestions made here https://github.com/facebook/jest/issues/9543 .  

I need a way to test code that uses require.resolve because currently, require.resolve starts at runner-ts/index.js whereas I need it to start from the tmp directory of my test.  Babel-core seems to use process.chdir with this "fake node-modules" approach, but I don't think that will  work for us.

File: plugins/runner-ts/src/index.ts [hunk 2]
```diff
@@ -61,7 +66,7 @@ export default defineRunner(tsRunnerOptions, async ctx => {
       const parentCompilerOptions = getCompilerOptionsRecursively(
         ctx.fs.existsSync(parentConfigPath)
           ? parentConfigPath
-          : require.resolve(currentConfigPath)
+          : require.resolve(extendsOption)
```
Comment (stephan-noel) [inline plugins/runner-ts/src/index.ts:65]: This was the only change I did for enabling npm module syntax.  I may be missing some other use case of having currentConfigPath here.  Can you verify?

Also, should this change be a separate PR?  It's pretty small.

File: plugins/runner-ts/src/index.ts [hunk 3]
```diff
@@ -95,47 +100,49 @@ export default defineRunner(tsRunnerOptions, async ctx => {
...
+      if (filesCache[denormalizePathForTSHost(fileName)]) {
         return true;
       }
       return ts.sys.fileExists(fileName);
     },
     getSourceFile(fileName) {
-      if (fileName.endsWith('.d.ts')) {
-        const stats = ctx.fs.statSync(fileName);
-        let file = filesCache[fileName];
+      const denormalziedFileName = denormalizePathForTSHost(fileName);
```
Comment (stephan-noel) [inline plugins/runner-ts/src/index.ts]: All of these changes for denormalizePathForTSHost had to do with the normalization done by the TS compiler.  Not sure if it should be a separate PR due to it touching the source code so much rather than only unit tests.

Also, I will need to manually test these changes some more since there was no bug that was manifesting, I'm not sure how this changes the behavior of the ts-runner on Windows.  I would expect that the ts-runner would fail on something so central to its functioning.

Also, will change this to just do path.normalize()

Comment (stephan-noel) [reply plugins/runner-ts/src/index.ts]: After debugging, in a test environment and production I've found that in production, there is no bug with respect to these changes in the TS host.   

In the test environment, fs is used directly which seems to be why the tests fail, in the production environment union-fs is used and it will somehow use fs behind the scenes, and successfully find the file on Windows even if it has a Unix path (also, the `registerInputFile` method initializes the cache with Unix files in production as well, ensuring correct cache hits).  

Also, I manually tested while debugging and it seems that union-fs works with my changes in which we pass a Windows path.

So in the end, it seems like `union-fs` makes normalization of these paths for the host unnecessary, (at least in the cases I've debugged so far).

So in the end we are with a problem of failing tests on Windows.  It seems that we have 2 options, either 1) normalize the paths like I am doing here or 2) use `union-fs` in our unit tests for the ts-runner.

@beshanoe What are your thoughts?

File: plugins/runner-ts/__tests__/default.test.ts (hunk 1 above)
Comment (stephan-noel) [reply plugins/runner-ts/__tests__/default.test.ts:89]: I was unable to get it done without using a custom jest resolver.   My resolver is based on the one mentioned in that issue except it's scoped so that module-resolution files for different tests don't affect each other.  It's maybe a bit much for one test, but maybe it can be useful to have.  What do you think?

File: plugins/runner-ts/src/index.ts (hunk 2 above)
Comment (beshanoe) [reply plugins/runner-ts/src/index.ts:65]: you're right, I think I've put currentConfigPath there by mistake and since I didn't add a test it got into the codebase :(

File: plugins/runner-ts/src/index.ts [hunk 4]
```diff
@@ -95,47 +96,49 @@ export default defineRunner(tsRunnerOptions, async ctx => {
     getCurrentDirectory: () => process.cwd(),
     getDefaultLibFileName: options => ts.getDefaultLibFilePath(options),
     fileExists: fileName => {
-      if (filesCache[fileName]) {
+      if (filesCache[Path.normalize(fileName)]) {
```
Comment (beshanoe) [inline plugins/runner-ts/src/index.ts:99]: so, Path.normalize is enough? No need for the normalize-path?

File: plugins/runner-postcss/__tests__/default.test.ts [hunk 5]
```diff
@@ -74,5 +95,38 @@ test('Resolves postcss @import', async () => {
...
+
+  if (os.platform() === 'win32') {
+    expect(hash).toBe('b2fa5e79e89879036728c19a038eafac');
+  } else {
+    expect(hash).toBe('8de170ce8b6d1c1aeb78d959f4faa0cf');
+  }
 });
+
+function extractHashes(
+  result: ExecuteRunner2Result,
```
Comment (beshanoe) [inline plugins/runner-postcss/__tests__/default.test.ts:110]: probably you could use `ReturnType<typeof executeRunner2>`

File: utils/fixture-helper/src/resolver.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import * as globby from 'globby';
+import Path = require('path');
```
Comment (beshanoe) [inline utils/fixture-helper/src/resolver.ts]: change to normal import from please

File: utils/fixture-helper/src/resolver.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  if (mapping[testDir]) {
+    const resolutionInFile = mapping[testDir][path];
+    actualResolution = resolutionInFile && Path.join(testDir, resolutionInFile);
+  }
+
+  return actualResolution || options.defaultResolver(path, options);
+}
+
+module.exports = resolver;
```
Comment (beshanoe) [inline utils/fixture-helper/src/resolver.ts:50]: and export too :)

File: plugins/runner-ts/src/index.ts (hunk 4 above)
Comment (stephan-noel) [reply plugins/runner-ts/src/index.ts:99]: From what I tested, Path.normalize seems to be correct, it's a bit confusing to me, because Path.normalize takes a path, and makes it platform-specific, whereas normalize-path takes a path, and conforms it to Unix (hence we can use them in jest tests on Windows to get a Unix path), yet they both suggest that they "normalize" the path.

File: utils/fixture-helper/src/resolver.ts (hunk 7 above)
Comment (stephan-noel) [reply utils/fixture-helper/src/resolver.ts:50]: It's pretty weird, but when I change to 

`export default resolver;`

In the compiled file, I have:

`exports.default = resolver;`

as expected, but when I run the tests I then get:

image

It's as if the resolver option of jest.config does not support `exports.default` syntax.

File: plugins/runner-ts/src/index.ts (hunk 2 above)
Comment (stephan-noel) [reply plugins/runner-ts/src/index.ts:65]: Ok nice, I was scared it may be support for some malformed path that I was destroying.  Kay, gonna resolve.

File: plugins/runner-postcss/__tests__/default.test.ts (hunk 5 above)
Comment (stephan-noel) [reply plugins/runner-postcss/__tests__/default.test.ts:110]: I tried but it didn't work as expected since executeRunner2 returns a Promise, and here I am using the wrapped value of the Promise.  I found some workaround using conditional type inference and it seems to work though it's a bit obscure `(Unpromise<T>)`.  Do you prefer this approach?

See https://github.com/Farfetch/garment/pull/9/files#diff-68fe2c93e76222c98020f0a5d9543adcR105

File: utils/fixture-helper/src/resolver.ts (hunk 6 above)
Comment (stephan-noel) [reply utils/fixture-helper/src/resolver.ts]: Done, what do you think?

File: plugins/runner-ts/src/index.ts (hunk 4 above)
Comment (beshanoe) [reply plugins/runner-ts/src/index.ts:99]: ah yeah, you're right

File: utils/fixture-helper/src/resolver.ts (hunk 7 above)
Comment (beshanoe) [reply utils/fixture-helper/src/resolver.ts:50]: oh I forgot it's used by jest directly, probably they only support commonjs, then it's ok you can leave it

File: plugins/runner-postcss/__tests__/default.test.ts (hunk 5 above)
Comment (beshanoe) [reply plugins/runner-postcss/__tests__/default.test.ts:110]: true, maybe then you can use unpromise to define the ExecuteRunner2Result type so it doesn't need to be kept in sync manually with what function returns, what do you think?

Comment (stephan-noel) [reply plugins/runner-postcss/__tests__/default.test.ts:110]: Hmm, I think I'm not understanding.  

I am doing `UnPromise<ReturnType<typeof executeRunner2>>` now and it works and I deleted the `ExecuteRunner2Result` so that type inference is used instead and it's working.  This way, there is nothing to keep in sync. 
 Do you want me to put the explicit type back in again?

Comment (beshanoe) [reply plugins/runner-postcss/__tests__/default.test.ts:110]: Sorry, I meant that in order to not define ad-hoc type helpers which are universal(like unpromise) in that particular test file, you could put them to where executeRunner2 is defined and use to create ExecuteRunner2Result type instead of defining it's detailed structure. But I think we can leave it as it's now as postcss test is the only place using ExecuteRunner2Result. But if you're foreseeing usage of this type in other tests, maybe it makes sense to put ExecuteRunner2Result next to executeRunner2 function and define it through the inference

Comment (stephan-noel) [reply plugins/runner-postcss/__tests__/default.test.ts:110]: Ahh yes, what you're saying makes perfect sense.  Also, by doing it that way, external TypeScript users of executeRunner2 can also now more easily define utility functions on the result as they see fit.  Changed it as suggested.  What do you think?

Also, I squashed and did a bit more manual testing of runner-ts to triple check and everything seems ok.  I think we're good to merge now.
